### ISSUE TITLE
VisIt can't connect client/server to sims when ssh tunneling is on.

### ISSUE THREAD

**Issue by aowen87** [aowen87/BAR#1893]
VisIt isn't successfully connecting client/server to sims when ssh tunneling is on. It works when ssh tunneling is off. However, when tunneling is on, the connection seems to be made but the type representations read in the sim contains all zeros. Maybe the socket bridge in vcl is to blame?

This is happening on the trunk but I think I have also seen it in released versions of VisIt.

-----------------------REDMINE MIGRATION-----------------------
This ticket was migrated from Redmine. As such, not all 
information was able to be captured in the transition. Below is
a complete record of the original redmine ticket.

Ticket number: 1193
Status: Resolved
Project: VisIt
Tracker: Bug
Priority: Normal
Subject: VisIt can't connect client/server to sims when ssh tunneling is on.
Assigned to: Brad Whitlock
Category: 
Target version: 2.7
Author: Brad Whitlock
Start: 09/06/2012
Due date: 
% Done: 100
Estimated time: 
Created: 09/06/2012 08:33 pm
Updated: 08/20/2013 03:13 pm
Likelihood: 3 - Occasional
Severity: 3 - Major Irritation
Found in version: 2.5.2
Impact: 
Expected Use: 
OS: All
Support Group: Any
Description:
 VisIt isn't successfully connecting client/server to sims when ssh tunneling is on. It works when ssh tunneling is off. However, when tunneling is on, the connection seems to be made but the type representations read in the sim contains all zeros. Maybe the socket bridge in vcl is to blame?

This is happening on the trunk but I think I have also seen it in released versions of VisIt.

Comments:
We've recently upgraded from VisIt 2.5.2 to 2.6.0, and are having sometrouble connecting to instrumented simulations. We're able to reproducethe problem with a simple test program (included below, with log from-debug 5 run), and it looks like a socket issue. We've already done all ofthe troubleshooting we could think of on our firewall settings, and hopethat maybe someone could help us here.Here's a walk through of what we see: We launch the simulation (compiledfrom the code below) on a linux node, and it starts normally. It creates asim2 file in ~/.visit/simulations with what I would expect are propersettings, the name of the node, the port, etc. And we have it set toblocking, so it waits for an incoming connection.At this point, we start VisIt 2.6.0 on Windows (with -debug 5 to get somedebug info) click open, select the linux host entry in the connectionwindow, which starts qssh (prompts me for my password), which launches themetadata server. I browse to the file location(~/.visit/simulation/*.sim2), and open the file.On the linux side, the libsim V2 code attempts to create 1 ÅreadÂ¹ and 3ÅwriteÂ¹ sockets, with the first two connections working, and the next twoconnections refused. It then uses the first (ÅreadÂ¹) socket to query theversion of the windows client. It's at this point that the recv functionfails (returns -1), but no error condition is generated yet. Shortlyafter, the code attempts to get the version of the windows client from thestill-empty buffer, and this evaluates to a blank version string. Whenit's compared to Â³2.6.0Â² local (linux) version string, it causes theversion check to fail and an exception to be generated. The resultingerror is written to the log Â³A.visit.5.vlogÂ² and the program exits with:Versions are (),2.6.0Exception: (IncompatibleVersionException)On the desktop side, the "compute engine launch progress" windows is stillwaiting, and has to be cancelled.Some questions we have are: what are the 1 'read' and '3' write socketsused for? Also, why would some of the sockets (first read and first write)connect, while the others have connection refused? It looks like even theconnected read socket has a problem, as the 'recv' called on it returns-1, leaving the zeroed-out buffer unaltered.The comments in the logs says that 9/10 of the issues are with firewalland host file config. What is the ideal setup in the /etc/hosts file?(should the hostname be listed with the 127.0.01 address? should itresolve itself to the actual IP of the first interface?Here is the sample code we used (condensed purposely):#include "VisItControlInterface_V2.h"#include <iostream>#include <stdlib.h>int main( int argc, char* argv[] ) {And the associated log:0visit -v 2.6 -engine -debuglaunch -noloopback -host node-01 -port 39879-key 89ea67c1a29831b2aa9dENGINE started. pid: 17154ParentProcess::Connect: Called with (numRead=1, numWrite=3, argc=14,argv={visit, -v, 2.6, -engine, -debuglaunch, -noloopback, -host, node-01,-port, 39879, -key, 89ea67c1a29831b2aa9d, -key, 89ea67c1a29831b2aa9d})ParentProcess::Connect: hostName = node-01ParentProcess::GetHostInfo: Calling gethostbyname("node-01")ParentProcess::Connect: port = 39879ParentProcess::Connect: securityKey = 89ea67c1a29831b2aa9dParentProcess::Connect: securityKey = 89ea67c1a29831b2aa9dParentProcess::Connect: Creating socketsParentProcess::Connect: Creating read socketsParentProcess::GetClientSocketDescriptor: Set up using port 39879ParentProcess::GetClientSocketDescriptor: Creating a socketParentProcess::GetClientSocketDescriptor: Setting socket optionsParentProcess::GetClientSocketDescriptor: Calling connect(If you see no messages after this one, VisIt was notable to connect to the client machine.  Nine times outof ten, this is a firewall issue on the client machine.It could also mean that VisIt was unable to resolve theIP address for the client machine.  You may need to verify the contents of/etc/hosts.)ParentProcess::GetClientSocketDescriptor: Connected socketParentProcess::Connect: Creating write socketsParentProcess::GetClientSocketDescriptor: Set up using port 39879ParentProcess::GetClientSocketDescriptor: Creating a socketParentProcess::GetClientSocketDescriptor: Setting socket optionsParentProcess::GetClientSocketDescriptor: Calling connect(If you see no messages after this one, VisIt was notable to connect to the client machine.  Nine times outof ten, this is a firewall issue on the client machine.It could also mean that VisIt was unable to resolve theIP address for the client machine.  You may need to verify the contents of/etc/hosts.)ParentProcess::GetClientSocketDescriptor: Connected socketParentProcess::GetClientSocketDescriptor: Set up using port 39879ParentProcess::GetClientSocketDescriptor: Creating a socketParentProcess::GetClientSocketDescriptor: Setting socket optionsParentProcess::GetClientSocketDescriptor: Calling connect(If you see no messages after this one, VisIt was notable to connect to the client machine.  Nine times outof ten, this is a firewall issue on the client machine.It could also mean that VisIt was unable to resolve theIP address for the client machine.  You may need to verify the contents of/etc/hosts.)ParentProcess::GetClientSocketDescriptor: Could not connect! (error=111:Connection refused)ParentProcess::GetClientSocketDescriptor: Set up using port 39879ParentProcess::GetClientSocketDescriptor: Creating a socketParentProcess::GetClientSocketDescriptor: Setting socket optionsParentProcess::GetClientSocketDescriptor: Calling connect(If you see no messages after this one, VisIt was notable to connect to the client machine.  Nine times outof ten, this is a firewall issue on the client machine.It could also mean that VisIt was unable to resolve theIP address for the client machine.  You may need to verify the contents of/etc/hosts.)ParentProcess::GetClientSocketDescriptor: Could not connect! (error=111:Connection refused)ParentProcess::Connect: Exchanging type representations.Versions are (),2.6.0Exception: (IncompatibleVersionException)/root/visit_2_6_0/visit2.6.0/src/common/comm/CommunicationHeader.C, line270: <The reason for the exception was not described>catch(IncompatibleVersionException)/root/visit_2_6_0/visit2.6.0/src/common/comm/ParentProcess.C:477Exception: (IncompatibleVersionException)/root/visit_2_6_0/visit2.6.0/src/common/comm/ParentProcess.C, line 510:<The reason for the exception was not described>catch(IncompatibleVersionException)/root/visit_2_6_0/visit2.6.0/src/engine/main/Engine.C:1354The engine has a different version than its client.
More information from William T Jones.I am using VisIt 2.6.2 to connect a client on my Linux desktop to aremote server through a login node via SSH tunneling.  While I can runVisIt in parallel (via PBS) to view files just fine, I am gettingwrapped around the axle with problems connecting to a libsiminstrumented simulation.  I cannot connect even to the simplest serialsimulation application which I found on this list (source included withthe logs).Basically: I start the test simulation on the remote machine; fire upVisIt on my desktop; Open a source; select the remote host; navigate tothe sim file in ".visit/simulations" of the remote host; wait forconnection and eventually Cancel.  The simple simulation execution lookslike:% ./VisItSimTestWaiting for connection from VisItVisIt did not connect to simulation<Enter> to exitqI have tried running the simple simulation on the login node as well ason a compute node of the server.  The attached "login_logs.tgz" containslogs from running simulation directly on the login server. The attached"node_logs.tgz" contains logs from running on a compute node.  Each hasthree directories: "client/" contains the logs from the client;"server/" contains the logs from the server; "sim/" contains the logsfrom the test simulation.I am sure this is a firewall or host table issue, but I cannot figure itout and am somewhat unclear as to the exact process involved.  I guessthat VisIt uses the `vcl` processes to launch things on the remoteserver, but I am unsure as to how the tunneling is supposed to work(e.g. Client tunnels to login node who forwards on to compute nodes).It appears that the simulation does not get the "-sshtunneling" CLO, butI don't know if it needs it by then.Any help would be appreciated.
I fixed the order of some calls in the viewer's code that launches the sim via vcl.
